pkg/catalog/catalog.go

```diff
@@ -1717,50 +1718,116 @@ func (c *Catalog) PrepareExpiredCommits(ctx context.Context, repositoryID string
 	return c.Store.SaveGarbageCollectionCommits(ctx, repository, previousRunID)
 }
 
-func createParquetFile(entries []DBEntry) (string, error) {
-	// TODO: Modify using a run ID generator (https://github.com/treeverse/lakeFS/issues/4469)
-	filename := uuid.New().String()
-	fw, err := local.NewLocalFileWriter(filename)
+type GCUncommittedMark struct {
+	BranchID graveler.BranchID `json:"branch"`
+	Key      graveler.Key      `json:"key"`
+}
+
+func (m *GCUncommittedMark) String() (string, error) {
+	var buf bytes.Buffer
+	encoder := base64.NewEncoder(base64.StdEncoding, &buf)
+	err := json.NewEncoder(encoder).Encode(m)
 	if err != nil {
 		return "", err
 	}
-	defer fw.Close()
-	pw, err := writer.NewParquetWriter(fw, new(DBEntry), int64(len(entries)))
+	result := buf.String()
+	return result, nil
+}
+
+type UncommittedParquetObject struct {
+	PhysicalAddress string `parquet:"name=physical_address, type=BYTE_ARRAY, convertedtype=UTF8, encoding=PLAIN_DICTIONARY"`
+	CreationDate    int64  `parquet:"name=creation_date, type=INT64, convertedtype=INT_64"`
+}
+
+func (c *Catalog) writeUncommittedLocal(ctx context.Context, repository *graveler.RepositoryRecord, fd *os.File, mark GCUncommittedMark) (*GCUncommittedMark, error) {
+	const parallelNum = 1                                                                        // Number of goroutines to handle marshaling of data
+	pw, err := writer.NewParquetWriterFromWriter(fd, new(UncommittedParquetObject), parallelNum) // TODO: Play with np count
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 	// Compression type
 	pw.CompressionType = parquet.CompressionCodec_GZIP
-	for _, entry := range entries {
-		if err = pw.Write(entry); err != nil {
-			return "", err
+
+	const GCMaxUncommittedFileSize = 20 * 1024 * 1024 // 20 MB
+	const periodicCheckSize = 100000
+	fileSizeExceeded := func() (bool, error) {
+		stat, err := fd.Stat()
+		if err != nil {
+			return false, err
 		}
+		return stat.Size() > GCMaxUncommittedFileSize, nil
 	}
-	if err = pw.WriteStop(); err != nil {
-		return "", err
-	}
-	_, err = fw.Seek(0, 0)
+
+	count := 0
+	bIt, err := c.Store.ListBranches(ctx, repository)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
-	return filename, nil
-}
-
-func markToString(mark *graveler.GCUncommittedMark) (*string, error) {
-	if mark == nil {
-		return nil, nil
+	defer bIt.Close()
+	bIt.SeekGE(mark.BranchID)
+	for bIt.Next() {
+		b := bIt.Value()
+		vIt, err := c.Store.ListStaging(ctx, b.Branch)
+		if err != nil {
+			return nil, err
+		}
+		if b.BranchID == mark.BranchID {
+			vIt.SeekGE(mark.Key)
+		}
+		eIt := NewValueToEntryIterator(vIt)
+		for eIt.Next() {
+			entry := eIt.Value()
+			count += 1
+			if count%periodicCheckSize == 0 {
+				err := pw.Flush(true)
+				if err != nil {
+					return nil, err
+				}
+				done, err := fileSizeExceeded()
+				if err != nil {
+					eIt.Close()
+					return nil, err
+				}
+				if done {
+					key := vIt.Value().Key
+					eIt.Close()
```

Comment:
The reason I'm using vIt - is to avoid a bug where the key is cast into path string (https://github.com/treeverse/lakeFS/issues/4554) - I don't believe this justifies a new iterator